src/utilities/createRouterPush.ts

```diff
@@ -1,15 +1,67 @@
-import { Resolved, Route, RouterPush, RouterPushOptions } from '@/types'
+import { Resolved, Route, RouteMethod, RouteMethods, Routes, isRouteMethodResponse } from '@/types'
+import { ExtractRoutePathParameters, RoutePaths } from '@/types/routePaths'
+import { flattenParentMatches } from '@/utilities/flattenParentMatches'
+import { isRecord } from '@/utilities/guards'
+import { normalizeRouteParams } from '@/utilities/normalizeRouteParams'
 import { RouterNavigation } from '@/utilities/routerNavigation'
+import { assembleUrl } from '@/utilities/urlAssembly'
+
+export type RouterPushOptions = {
+  query?: Record<string, unknown>,
+  replace?: boolean,
+}
+
+export type RouteWithParams<
+  TRoutes extends Routes,
+  TRoutePath extends string
+> = {
+  route: TRoutePath & RoutePaths<TRoutes>,
+  params: ExtractRoutePathParameters<RouteMethods<TRoutes>, TRoutePath>,
+}
+
+export type RouterPush<
+  TRoutes extends Routes
+> = <
+  TRoutePath extends string
+>(source: string | RouteWithParams<TRoutes, TRoutePath> | ReturnType<RouteMethod>, options?: RouterPushOptions) => Promise<void>
 
 type RouterPushContext = {
   navigation: RouterNavigation,
   resolved: Resolved<Route>[],
 }
 
-export function createRouterPush({ navigation }: RouterPushContext): RouterPush {
-  const push: RouterPush = (url, options?: RouterPushOptions) => {
-    return navigation.update(url, options)
+export function createRouterPush<
+  TRoutes extends Routes
+>({ navigation, resolved }: RouterPushContext): RouterPush<TRoutes> {
+  return (source, options) => {
+    if (typeof source === 'string') {
+      return navigation.update(source, options)
+    }
+
+    if (isRouteMethodResponse(source)) {
+      return navigation.update(source.url, options)
+    }
+
+    if (isRecord(source)) {
+      const match = resolved.find((resolvedRoute) => flattenParentMatches(resolvedRoute) === source.route)
+
+      if (!match) {
+        throw `No route found: "${String(source)}"`
+      }
+
+      const params = get<Record<string, unknown>>(source, 'params')
```

Comment:
I've 100% written a type like this when I got annoyed by Typescript but ultimately I think we're just hiding the `as Record<string, unknown>` in a method. I personally would drop the excess in favor of just casting here

```suggestion
      const params = source.params as Record<string, unknown>
```

Reply:
never mind, I see now how this avoid TS checking at all. Unfortunate but makes sense now